### ISSUE TITLE
[SPIKE] Deep Links 2.0

### ISSUE THREAD

**Issue by jesmrec** [owncloud/android#4068]
Research about handling deep links

Research 1.0: https://github.com/owncloud/enterprise/issues/4970#issuecomment-1042967713

What's expected on Deep Links:

user clicks on a link in the device (no matter where: mail, web...):

1. App is opened
2. Account is selected, if more than one
3. Browse through folder structure to reach the pointed item. In case it's not discovered, all the structure must be discovered on the road.

What should be researched:

- How brandable could the link be. Basic link is something like `https://demo.owncloud.com/f/7`. Prefix and URL to be brandable, so that we can open:  `mycompany://mycompany.thebest.com/index.php/f/13` with same effect. Scheme, host and preffix could have different values. (here there are some Android version regards, i guess). Desirable: use as default `owncloud://` scheme.

- Check if this [first iteration](https://github.com/owncloud/android/pull/3598) is still useful. In there, links to already discovered items are open. Branding options for host and preffix also there.

- Which information to be sent to backend team to make this posible. Automatic verifications could be involved (check [this](https://github.com/owncloud/enterprise/issues/4970#issuecomment-1044159479))

- Logic to develop, check feasibility and estimation (maybe the longest part)
   - Solve link to get the path to the item: https://github.com/owncloud/enterprise/issues/4970#issuecomment-1040540303
   - Browse/discover to open item (file or folder). Take in account the browsability.

- Does make sense to split up in different parts as in the 1st try?:
   -  1) Open discovered files and folders in single account
   - 2) Open non-discovered in single account
   - 3) Open anything in multiaccount

- If time allows, do a quick proof of concept of the easiest case

Any other idea ... always welcome.

**Comment by jesmrec** [comment]
@michaelstingl one questions about expected behaviour: working in both oC10 / oCIS? in that case, same link format in both?

In case of oCIS, we have to add another step: check which space/drive does contain the folder.

**Comment by michaelstingl** [comment]
> > oCIS focus. So yes, we'll have to deal with Spaces. But meta endpoint should provide this information:
> 
> this will translate the file id into a path to reach the file. But, does the link itself look like in oC10? (i hope so):
> 
> `<scheme>://<url>/f/<some_id>`
> 
> i did not find this kind of links in oCIS web, only the public links with the `/s/` prefix.

Yes, needs more concepting with oCIS + web

/cc @tbsbdr 

Currently you can only use the link of a folder from the browser address bar:

`https://ocis.ocis-traefik.latest.owncloud.works/files/spaces/personal/einstein/New%20folder?fileId=96fd1fa6-2d2a-4701-be6e-7d15a82ecd95%244c510ada-c86b-4815-8820-42cdf82c3d51%2174cb2d7f-5c0f-4af9-8903-d9984294ab8b&items-per-page=100&view-mode=resource-table&tiles-size=1&sort-dir=asc&sort-by=name`

**Comment by michaelstingl** [comment]
## oC10 link / fileid resolve

```
% curl -sI -u "test:test" "https://demo.owncloud.com/f/9" | grep "location"
location: /apps/files/?dir=/Photos
webdav-location: /remote.php/dav/files/test/Photos
```

- https://github.com/owncloud/core/pull/30383

## oCIS link / fileid resolve

`/remote.php/dav/meta/` could be used to get more information.

```
% curl -s 'https://ocis.ocis-traefik.latest.owncloud.works/remote.php/dav/meta/427a5e74-6921-4768-8d53-cd5e8b30f5b0%24534bb038-6f9d-4093-946f-133be61fa4e7!a130bbe3-5681-4193-9ab0-2a714934722a' \
-X 'PROPFIND' \
-H 'Authorization: Bearer REDACTED' \
-H 'Depth: 0' \
--data-binary $'<?xml version="1.0"?>\n<d:propfind  xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">\n  <d:prop>\n    <oc:meta-path-for-user />\n  </d:prop>\n</d:propfind>' | xmllint --format -

<?xml version="1.0"?>
<d:multistatus xmlns:s="http://sabredav.org/ns" xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/remote.php/dav/meta/427a5e74-6921-4768-8d53-cd5e8b30f5b0$534bb038-6f9d-4093-946f-133be61fa4e7%21a130bbe3-5681-4193-9ab0-2a714934722a/</d:href>
    <d:propstat>
      <d:prop>
        <oc:meta-path-for-user>/Hacker-Songs</oc:meta-path-for-user>
        <oc:id>427a5e74-6921-4768-8d53-cd5e8b30f5b0$534bb038-6f9d-4093-946f-133be61fa4e7!a130bbe3-5681-4193-9ab0-2a714934722a</oc:id>
        <oc:fileid>427a5e74-6921-4768-8d53-cd5e8b30f5b0$534bb038-6f9d-4093-946f-133be61fa4e7!a130bbe3-5681-4193-9ab0-2a714934722a</oc:fileid>
        <oc:spaceid>427a5e74-6921-4768-8d53-cd5e8b30f5b0</oc:spaceid>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

- https://doc.owncloud.com/server/developer_manual/webdav_api/meta.html

**Comment by michaelstingl** [comment]
@jesmrec @JuancaG05 please provide feedback/ideas if anything else needed from the backend

**Comment by manuelplazaspalacio** [comment]
I continue checking the possibilities tu open an url with a custom scheme in Android. I create an app to open a page with a "oc" scheme inside a webview. If you use the WebView as is, when you click on the link, the WebView displays an error saying it doesn't recognize that scheme. Later, I added the code shown below to handle deep links and it opens the ownCloud app correctly. So, there is a possibility that if you try to open the deep link from a WebView within an app, it may not open. On the other hand, I have tested it on the 9 most commonly used Android browsers and it works fine there.

**Comment by michaelstingl** [comment]
> On the other hand, I have tested it on the 9 most commonly used Android browsers and it works fine there.

This sounds great 👍

**Comment by manuelplazaspalacio** [comment]
I have been running tests, and I believe that with this, I can now consider the spike completed. I have obtained the file path through the request to "/remote.php/dav/meta/" with the file's ID. I have updated the containing folders of the file, and then I navigate to the file and display its details.

In this video, it shows how it navigate to the file's details automatically using the deep link, then I view the image and manually navigate back to the initial folder:


https://github.com/owncloud/android/assets/24573727/78de7034-f14c-461a-bd44-4b8ad19df9e7


In this other video, I demonstrate how to navigate automatically to a folder using the deep link and then manually navigate back to the initial folder:

https://github.com/owncloud/android/assets/24573727/e015c1b7-827e-4f01-8674-9b3adde77154
